File: app/src/main/java/de/roland_illig/phantomgo/Board.java
package de.roland_illig.phantomgo;

import android.support.annotation.VisibleForTesting;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

public class Board {
    private final int size;

    private final Player[][] pieces;
    private final int[] captured = new int[2];
    private Player turn;
    private Point prevMove;
    private boolean passed;
    private boolean gameOver;

    private RefereeListener refereeListener;
    private UpdateListener updateListener;

    public Board(int size) {
        this.size = size;
        pieces = new Player[size][size];
        reset();
    }

    public void reset() {
        turn = Player.BLACK;
        for (Player[] row : pieces) {
            Arrays.fill(row, null);
        }
        prevMove = new Point(-1, -1);
        Arrays.fill(captured, 0);
        passed = false;
        gameOver = false;

        if (updateListener != null) {
            updateListener.onUpdate();
        }
    }

    public Board copy() {
        Board copy = new Board(size);
        for (int x = 0; x < size; x++) {
            System.arraycopy(pieces[x], 0, copy.pieces[x], 0, size);
        }
        System.arraycopy(captured, 0, copy.captured, 0, captured.length);
        copy.turn = turn;
        copy.prevMove = prevMove;
        return copy;
    }

    public Player get(int x, int y) {
        return pieces[x][y];
    }

    public int getCaptured(Player player) {
        return captured[player.ordinal()];
    }

    private int tryCapture(int x, int y, Player turn) {
        if (0 <= x && x < size && 0 <= y && y < size) {
            if (getLiberties(x, y, turn) == 0) {
                return capture(x, y, turn);
            }
        }
        return 0;
    }

    private int getLiberties(int x, int y, Player color) {
        return get(x, y) == color ? getLiberties(x, y) : -1;
    }

    @VisibleForTesting
    int getLiberties(int x, int y) {
        LibertiesCounter counter = new LibertiesCounter(get(x, y));
        counter.count(x, y);
        return counter.liberties.size();
    }

    private int capture(int x, int y, Player color) {
        int captured = 1;
        pieces[x][y] = null;
        if (x > 0 && pieces[x - 1][y] == color)
            captured += capture(x - 1, y, color);
        if (x < size - 1 && pieces[x + 1][y] == color)
            captured += capture(x + 1, y, color);
        if (y > 0 && pieces[x][y - 1] == color)
            captured += capture(x, y - 1, color);
        if (y < size - 1 && pieces[x][y + 1] == color)
            captured += capture(x, y + 1, color);
        return captured;
    }

    public void setRefereeListener(RefereeListener refereeListener) {
        this.refereeListener = refereeListener;
    }

    public void setUpdateListener(UpdateListener updateListener) {
        this.updateListener = updateListener;
    }

    public void set(int x, int y, Player color) {
        pieces[x][y] = color;
        if (updateListener != null) {
            updateListener.onUpdate();
        }
    }

    public void setTurn(Player turn) {
        this.turn = turn;
    }

    public interface RefereeListener {
        void onReferee(RefereeResult result);
    }

    public interface UpdateListener {
        void onUpdate();
    }

    private class LibertiesCounter {
        final Player color;
        final Set<Point> todo = new HashSet<>();
        final Set<Point> done = new HashSet<>();
        final Set<Point> liberties = new HashSet<>();

        LibertiesCounter(Player color) {
            color.getClass();
            this.color = color;
        }

        void count(int x, int y) {
            todo.add(new Point(x, y));

            while (!todo.isEmpty()) {
                Iterator<Point> it = todo.iterator();
                Point point = it.next();
                it.remove();
                done.add(point);

                int px = point.x;
                int py = point.y;
                if (px > 0) {
                    countInternal(px - 1, py);
                }
                if (py > 0) {
                    countInternal(px, py - 1);
                }
                if (px < size - 1) {
                    countInternal(px + 1, py);
                }
                if (py < size - 1) {
                    countInternal(px, py + 1);
                }
            }
        }

        private void countInternal(int nx, int ny) {
            Point np = new Point(nx, ny);
            Player neighbor = get(nx, ny);
            if (neighbor == color && !done.contains(np)) {
                todo.add(np);
            }
            if (neighbor == null) {
                liberties.add(np);
            }
        }
    }

    public int getSize() {
        return size;
    }

    public RefereeResult play(int x, int y) {
        RefereeResult result = playInternal(x, y);
        if (refereeListener != null) {
            refereeListener.onReferee(result);
        }
        if (result.invalidReason == null && updateListener != null) {
            updateListener.onUpdate();
        }
        return result;
    }

    private RefereeResult playInternal(int x, int y) {
        if (gameOver) {
            throw new IllegalStateException("GameOver");
        }
        Player other = turn.other();

        if (get(x, y) == turn) {
            return RefereeResult.ownStone();
        }
        if (get(x, y) == other) {
            return RefereeResult.otherStone();
        }

        boolean selfLeftBefore = x > 0 && getLiberties(x - 1, y, turn) == 1;
        boolean selfAboveBefore = y > 0 && getLiberties(x, y - 1, turn) == 1;
        boolean selfRightBefore = x + 1 < size && getLiberties(x + 1, y, turn) == 1;
        boolean selfBelowBefore = y + 1 < size && getLiberties(x, y + 1, turn) == 1;
        boolean selfAtariBefore = selfLeftBefore || selfAboveBefore || selfRightBefore || selfBelowBefore;

        int leftBefore = x > 0 ? getLiberties(x - 1, y, other) : 0;
        int aboveBefore = y > 0 ? getLiberties(x, y - 1, other) : 0;
        int rightBefore = x + 1 < size ? getLiberties(x + 1, y, other) : 0;
        int belowBefore = y + 1 < size ? getLiberties(x, y + 1, other) : 0;

        int dx = x - prevMove.x;
        int dy = y - prevMove.y;
        if (dx * dx + dy * dy == 1) {
            if (getLiberties(prevMove.x, prevMove.y, other) == 1) {
                return RefereeResult.ko();
            }
        }

        pieces[x][y] = turn;
        boolean undo = true;
        try {
            boolean capturesSomething = leftBefore == 1 || aboveBefore == 1 || rightBefore == 1 || belowBefore == 1;
            if (!capturesSomething && getLiberties(x, y) == 0) {
                return RefereeResult.suicide();
            }

            int left = x > 0 ? getLiberties(x - 1, y, other) : 0;
            int above = y > 0 ? getLiberties(x, y - 1, other) : 0;
            int right = x + 1 < size ? getLiberties(x + 1, y, other) : 0;
            int below = y + 1 < size ? getLiberties(x, y + 1, other) : 0;

            boolean atari = leftBefore > 1 && left == 1
                    || aboveBefore > 1 && above == 1
                    || rightBefore > 1 && right == 1
                    || belowBefore > 1 && below == 1;

            int capturedStones = tryCapture(x - 1, y, other)
                    + tryCapture(x, y - 1, other)
                    + tryCapture(x + 1, y, other)
                    + tryCapture(x, y + 1, other);

            boolean selfAtari = getLiberties(x, y, turn) == 1 && !selfAtariBefore;

            captured[turn.ordinal()] += capturedStones;
            prevMove = capturedStones == 1 && selfAtari ? new Point(x, y) : new Point(-1, -1);
            turn = turn.other();
            undo = false;

            return RefereeResult.ok(atari, selfAtari, capturedStones);
        } finally {
            if (undo) {
                pieces[x][y] = null;
            }
        }
    }

    public RefereeResult pass() {
        if (gameOver) {
            throw new IllegalStateException("GameOver");
        }
        gameOver = passed;
        passed = true;
        turn = turn.other();
        if (updateListener != null) {
            updateListener.onUpdate();
        }
        return RefereeResult.pass();
    }

    public boolean isGameOver() {
        return gameOver;
    }

    public Player getTurn() {
        return turn;
    }

    @VisibleForTesting
    public void setup(String... rows) {
        for (int y = 0; y < size; y++) {
            for (int x = 0; x < size; x++) {
                char ch = rows[y].charAt(2 * x);
                pieces[x][y] = parseChar(ch);
            }
        }
    }

    private static Player parseChar(char ch) {
        if (ch == 'W') return Player.WHITE;
        if (ch == 'B') return Player.BLACK;
        if (ch == '.') return null;
        throw new IllegalStateException();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int y = 0; y < size; y++) {
            for (int x = 0; x < size; x++) {
                Player player = get(x, y);
                sb.append(player != null ? "BW".charAt(player.ordinal()) : '.');
                sb.append(x == size - 1 ? "\n" : " ");
            }
        }
        return sb.toString();
    }

    private static class Point {
        private final int x;
        private final int y;

        private Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            Point point = (Point) obj;
            return x == point.x && y == point.y;
        }

        @Override
        public int hashCode() {
            return 31 * x + y;
        }
    }
}

File: app/src/main/java/de/roland_illig/phantomgo/Board.kt
package de.roland_illig.phantomgo

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

open class Board(val size: Int) : java.io.Serializable {

    var rules = Rules.Normal

    private val pieces = Array(size) { Array<Player?>(size) { null } }
    private val captured = IntArray(2)
    var turn = Player.BLACK
    private var koMove: Intersection? = null
    private var passed = 0
    var gameOver = false; private set
    val started get() = lastMove != null || passed > 0
    var lastMove: Intersection? = null; private set

    fun copy(): Board = ByteArrayOutputStream().let {
        ObjectOutputStream(it).writeObject(this)
        return ObjectInputStream(ByteArrayInputStream(it.toByteArray())).readObject() as Board
    }

    operator fun get(pos: Intersection) = pieces[pos.x][pos.y]

    operator fun set(pos: Intersection, color: Player?) {
        pieces[pos.x][pos.y] = color
    }

    override fun toString() = toStringLines().joinToString("") { it + "\n" }

    fun toStringLines(): List<String> {
        val lines = mutableListOf<String>()
        val sb = StringBuilder()
        for (y in 0 until size) {
            for (x in 0 until size) {
                sb.append(".XO"[1 + (pieces[x][y]?.ordinal ?: -1)])
                if (x < size - 1) sb.append(" ")
            }
            lines += "$sb"
            sb.setLength(0)
        }
        return lines
    }

    fun getCaptured(player: Player) = captured[player.ordinal]

    fun edit(pos: Intersection, color: Player) {
        turn = color
        val result = play(pos)
        if (result !is RefereeResult.Invalid) return
        val newColor = if (result is RefereeResult.OwnStone) null else color
        set(pos, newColor)
    }

    operator fun set(x: Int, y: Int, color: Player?) {
        pieces[x][y] = color
    }

    fun pass(): RefereeResult {
        check(!gameOver) { "GameOver" }

        lastMove = null
        passed++
        gameOver = passed >= 2
        turn = turn.other()
        return RefereeResult.Pass
    }

    fun play(pos: Intersection): RefereeResult {
        check(!gameOver) { "GameOver" }

        val turn = turn
        val other = turn.other()

        if (this[pos] == turn) return RefereeResult.OwnStone
        if (this[pos] == other) return RefereeResult.OtherStone

        val neighbors = neighbors(pos)
        val koMove = koMove
        if (koMove != null && koMove in neighbors && isAtari(koMove, other)) {
            return RefereeResult.Ko
        }

        val prevBoard = copy()

        this[pos] = turn
        if (neighbors.none { prevBoard.isAtari(it, other) } && getLiberties(pos) == 0) {
            this[pos] = null
            return RefereeResult.Suicide
        }

        val placed = mutableListOf(pos)
        val captured = mutableListOf<Intersection>()
        moveMagnetic(pos, placed)
        val atari = captureOther(placed, other, captured, prevBoard)
        val selfAtari = captureTurn(placed, turn, captured, prevBoard)

        this.captured[turn.ordinal] += captured.size
        this.koMove = if (captured.size == 1 && selfAtari) pos else null
        this.turn = other
        lastMove = pos
        passed = 0

        return RefereeResult.Ok(atari, selfAtari, captured.toList())
    }

    private fun moveMagnetic(pos: Intersection, placed: MutableList<Intersection>) {
        if (rules != Rules.Magnetic) return

        data class Direction(val x: Int, val y: Int)

        operator fun Intersection.plus(dir: Direction) =
            Intersection(this.x + dir.x, this.y + dir.y)

        operator fun Intersection.minus(dir: Direction) =
            Intersection(this.x - dir.x, this.y - dir.y)

        fun move(dx: Int, dy: Int) {
            val dir = Direction(dx, dy)
            var from = pos + dir
            while (from.ok() && this[from] == null) from += dir
            if (!from.ok()) return

            val to = if (this[from] == turn) {
                var to = from + dir
                while (to.ok() && this[to] == null) to += dir
                to - dir
            } else {
                pos + dir
            }

            if (to == from) return
            this[to] = this[from]
            this[from] = null
            placed += to
        }

        move(0, -1)
        move(-1, 0)
        move(+1, 0)
        move(0, +1)
    }

    private fun captureOther(
        placed: List<Intersection>,
        other: Player,
        captured: MutableList<Intersection>,
        prevBoard: Board
    ): Boolean {
        var atari = false
        for (p in placed) {
            for (n in neighbors(p)) {
                maybeCapture(n, other, captured)
                if (isAtari(n, other) && !prevBoard.isAtari(n, other)) atari = true
            }
        }
        return atari
    }

    private fun captureTurn(
        placed: List<Intersection>,
        turn: Player,
        captured: MutableList<Intersection>,
        prevBoard: Board
    ): Boolean {
        var selfAtari = false
        for (p in placed) {
            maybeCapture(p, turn, captured)
            if (!isAtari(p, turn)) continue
            if (prevBoard.neighbors(p).any { prevBoard.isAtari(it, turn) }) continue
            selfAtari = true
        }
        return selfAtari
    }

    private fun isAtari(pos: Intersection, color: Player) =
        pos.ok() && get(pos) == color && getLiberties(pos) == 1

    private fun maybeCapture(
        pos: Intersection,
        color: Player,
        captured: MutableList<Intersection>
    ) {
        if (!pos.ok()) return
        if (!(get(pos) == color && getLiberties(pos) == 0)) return

        fun takeOut(pos: Intersection) {
            if (this[pos] != color) return

            this[pos] = null
            captured += pos
            for (n in neighbors(pos)) if (n.ok()) takeOut(n)
        }

        takeOut(pos)
    }

    fun getLiberties(pos: Intersection): Int {
        val color = this[pos]!!
        val todo = mutableSetOf<Intersection>()
        val done = mutableSetOf<Intersection>()
        val liberties = mutableSetOf<Intersection>()

        fun countInternal(np: Intersection) {
            val neighbor = this[np]
            if (neighbor == color && !done.contains(np)) todo.add(np)
            if (neighbor == null) liberties.add(np)
        }

        todo.add(pos)

        while (todo.isNotEmpty()) {
            val it = todo.iterator()
            val point = it.next()
            it.remove()
            done.add(point)

            for (n in neighbors(point)) countInternal(n)
        }

        return liberties.size
    }

    private fun neighbors(pos: Intersection): List<Intersection> {
        val max = size - 1
        val (x, y) = pos
        return if (rules == Rules.Toroidal) {
            listOf(
                Intersection(x, if (y > 0) y - 1 else max),
                Intersection(if (x > 0) x - 1 else max, y),
                Intersection(if (x < max) x + 1 else 0, y),
                Intersection(x, if (y < max) y + 1 else 0)
            )
        } else {
            val neighbors = mutableListOf<Intersection>()
            if (y > 0) neighbors += Intersection(x, y - 1)
            if (x > 0) neighbors += Intersection(x - 1, y)
            if (x < max) neighbors += Intersection(x + 1, y)
            if (y < max) neighbors += Intersection(x, y + 1)
            neighbors
        }
    }

    private fun Intersection.ok() = this.x in 0 until size && this.y in 0 until size
}
